File: codes/ruby/chapter09/answer_A67.rb
# Union-Find Tree
class UnionFind
  # Create a Union-Find Tree with n vertices
  def initialize(n)
    @n = n
    @par = Array.new(n + 1, -1) # Initially, there are no parents
    @size = Array.new(n + 1, 1) # Initially, the number of vertices in the group is 1
  end

  # Returns the root of vertex x
  def root(x)
    # Continue to move one step forward (parent) until there is no parent (i.e., reach the root)
    while @par[x] != -1
      x = @par[x]
    end
    x
  end

  # Unite elements u and v
  def unite(u, v)
    rootu = root(u)
    rootv = root(v)
    if rootu != rootv
      # Process only when u and v are in different groups
      if @size[rootu] < @size[rootv]
        @par[rootu] = rootv
        @size[rootv] += @size[rootu]
      else
        @par[rootv] = rootu
        @size[rootu] += @size[rootv]
      end
    end
  end

  # Returns whether elements u and v are in the same group
  def same(u, v)
    root(u) == root(v)
  end
end

# Input
n, m = gets.split.map(&:to_i)
edges = Array.new(m) { gets.split.map(&:to_i) }

# Sort edges by length in ascending order
edges.sort_by! { |x| x[2] }

# Find the minimum spanning tree
uf = UnionFind.new(n)
answer = 0
edges.each do |a, b, c|
  unless uf.same(a, b)
    uf.unite(a, b)
    answer += c
  end
end

# Output the answer
puts answer

File: codes/go/chapter09/answer_A67.go
package main

import (
	"fmt"
	"sort"
)

type UnionFind struct {
	parent []int
	size   []int
}

func newUnionFind(n int) *UnionFind {
	parent := make([]int, n+1)
	size := make([]int, n+1)
	for i := 1; i <= n; i++ {
		parent[i] = -1
		size[i] = 1
	}
	return &UnionFind{parent, size}
}

func (uf *UnionFind) root(x int) int {
	for uf.parent[x] != -1 {
		x = uf.parent[x]
	}
	return x
}

func (uf *UnionFind) unite(u, v int) {
	rootU := uf.root(u)
	rootV := uf.root(v)
	if rootU == rootV {
		return
	}

	if uf.size[rootU] < uf.size[rootV] {
		uf.parent[rootU] = rootV
		uf.size[rootV] += uf.size[rootU]
	} else {
		uf.parent[rootV] = rootU
		uf.size[rootU] += uf.size[rootV]
	}
}

func (uf *UnionFind) same(u, v int) bool {
	return uf.root(u) == uf.root(v)
}

type Edge struct {
	a, b, c, idx int
}

func main() {
	var n, m int
	fmt.Scanf("%d %d", &n, &m)

	edges := make([]Edge, m)
	for i := 0; i < m; i++ {
		var a, b, c int
		fmt.Scanf("%d %d %d", &a, &b, &c)
		edges[i] = Edge{a, b, c, i + 1}
	}

	// 辺の長さの小さい順にソート
	sort.Slice(edges, func(i, j int) bool {
		return edges[i].c < edges[j].c
	})

	answer := 0
	uf := newUnionFind(n)

	for _, edge := range edges {
		if !uf.same(edge.a, edge.b) {
			uf.unite(edge.a, edge.b)
			answer += edge.c
		}
	}

	fmt.Println(answer)
}
